### ISSUE TITLE
Question about state/variables

### ISSUE THREAD

**Issue by Madd0g** [nikkorn/mistreevous#97]
Cool project!

Have you considered having variables support in the DSL? A context that the DSL could reference and the actions could update?

I'm kinda new to this, so maybe I'm missing something about state management, but it feels like being able to pass an object with data and being able to reference that from the DSL can be very useful.

Are any other examples or guides of complex state management, like:

- what if steps need outputs from previous steps (partially the context thing addresses this)
- dealing with failures
- assigning ids or extra props to nodes to know which node is which (for observability, I haven't checked how the visualizer does it yet)

Will appreciate any feedback!

**Comment by nikkorn** [comment]
Thanks very much @Madd0g ! 👍 

Adding support for variables is something I've never really considered as I had designed this library in such a way that all state is the responsibility of the agent to manage, as opposed to other libraries out there which do handle some state for the agent, mainly in the form of a blackboard. Allowing the agent to manage its own state has its benefits, but I do agree that it would be very convenient to be able to pass variables to agent functions and not have to use literals everywhere. How we manage those variables, where they are stored/defined, and how we reference these in our mdsl/json is the real question though, do you have any ideas on how to maybe approach this?

Some initial ideas:

1.  Allow users to reference agent properties when defining action/condition/guard/callback function arguments. (easier)
2.  Each tree instance has a blackboard object (the initial state of this blackboard could be an optional BehaviourTree constructor parameter) and any conditions/actions can get/set values on the blackboard via methods on the tree instance. Any properties on the blackboard can be referenced when defining action/condition/guard/callback function arguments.

**what if steps need outputs from previous steps (partially the context thing addresses this)**
This would be a case of setting some state on the agent as part of one step and then simply using this state in subsequent steps. E.g. You may have a FindPath action which does some path finding and sets an 'activePath' property on the agent, and the next action to be visited may be FollowActivePath which just uses that 'activePath' property to update the position of the agent.
I havent looked too much into how other libraries which implement a blackboard would expect users to distribute state between the agent and the tree intance. I will have a read up on that.

**dealing with failures**
This is handled by the user by means of:
- Returning the failure state from actions.
- Conditional checks.
- Aborting running nodes via guards.

**Assigning ids or extra props to nodes**
This would only be something that is useful for debugging/visualisations. While it would be relatively easy to allow users to assign identifers and any custom data to a node definition in json it would be considerable more work to implement this in mdsl so i'd be inclined not to do it.

Do these answers help or were you looking for more general information about how these things work with behaviour trees in general and weren't being specific to this library?

Thanks again for the feedback, I really appreciate it, it's always great to get a fresh perspective.

**Comment by Madd0g** [comment]
> Do these answers help or were you looking for more general information about how these things work with behaviour trees in general and weren't being specific to this library?

I'm a beginner in both the general and the specific, so everything definitely helps.

Thanks for replying, I'll respond about the variables (that's what I care about the most) and maybe come back later for a full response

> but I do agree that it would be very convenient to be able to pass variables to agent functions and not have to use literals everywhere.

That's what I'm aiming for, not just to eliminate the hardcoded literals but also give the DSL more flexibility/control, with variables the DSL (not just the agent) can be more dynamic by referencing things from this blackboard object (not just use static literals, but also objects?)

**Where they are stored** - that's a problem due to the lack of support in JSON for anything other than simple types. Maybe variable-access could be wrapped in with a special wrapper `{{obj.value}}` or something in the JSON. But is there another place where this causes concerns?

```
root {
    condition [HasItem, "gold", 500]
}
```

Could be:

```
root {
    condition [HasItem, targetOre, minQuantityForSale]
}
```

And:

```json
{
    "type": "root",
    "child": {
        "type": "condition",
        "call": "HasItem",
        "args": [
            "{{ctx.targetOre}}",
           "{{ctx.minQuantityForSale}}"
        ]
    }
}
```

**Where's the blackboard?** 

Maybe it deserves a special object under the agent and not the agent itself - it's just allows for more interesting behaviors and limits scope.

One thing I did before in a similar situation is to use a Proxy object to resolve variable access upwards in whatever chain the code was in. So if you want to have separate blackboards - the inner one could transparently reach into the blackboard of the parent. This is _kinda black magic_ and I'm not sure it's necessary here, but it allows for interesting compositions.

I (briefly) looked into other BT libraries and didn't really see anything with a DSL, I think in a code-only library variable/object access is just easier/different?

Thanks for considering this!

**Comment by nikkorn** [comment]
@Madd0g Alright, I've had a good think about this. My only real issue with this was adding a blackboard. I feel like it doesn't match well with the stateless design of this behaviour tree implementation, and even if I did add it I cant think of a nice api for it.

After eventually coming across a really good paper on tree parameterisation (https://people.cs.umass.edu/~fmgarcia/Papers/Parameterizing%20Behavior%20Trees.pdf) it's reinforced my decision to avoid blackboards (for now at least) but to still add support for variables, which will just be references to either agent properties or subtree parameters.

I think that is a nice middle-ground in allowing us to be more flexible with state but not have to adopt a full blackboard and the issues that having one can cause.

I like your suggestions about how these variables could be defined in json and mdsl so I will use that (excluding the 'ctx.' qualifier for the json one as I think we don't need it)

I was thinking that it might also be a good idea to add a few built-in functions (IsTrue, IsFalse, IsSet, IsNotSet, IsEqual, IsNotEqual etc.) to reference in place of agent or globally registered functions, just to make it a little nicer to work with these variables. Some examples:

```
root {
    sequence {
        condition [IsTrue, isAlive]
        action [Say, "I'm Alive!"]
    }
}
```

```
root {
    sequence {
        condition [IsSet, activeTarget]
        action [MoveTowardsActiveTarget]
    }
}
```

```
root {
    sequence {
        condition [IsEqual, equippedWeapon, "Hammer"]
        action [Smash]
    }
}
```

I also want to add the ability to parameterise subtrees which I think will really help with their reusability, especially any globally registered subtrees. For example:

```
//  Globally registered as the 'DropItem' subtree
root [itemType, dropLocation] {
     sequence {
        condition [HasItem, itemType]
        action [WalkToLocation, dropLocation]
        action [RemoveInventoryItem, itemType]
    }
}


// Character tree definition
root {
    branch [DropItem, "longsword", "trashcan"]
}
```

What do you think? Thanks very much :)

**Comment by Madd0g** [comment]
I think your examples are great and this type of variable use is exactly what I was going for, I feel that this little touch of additional dynamicity in the DSL really adds a lot.

So, the DSL variables will be able to reference any property on the Agent?

So, I could assign them freely? From all these:

1. when creating the agent
2. at any point from agent functions
3. create a smart JS getter

How about keyed access `something.other` or `myArr[0]`?

Thanks!

**Comment by nikkorn** [comment]
Yeah I definitely think it will make definitions much more dynamic and really help with subtree reuse.

So yes, mdsl variables (and references via "{{someAgentProperty}}" in json) will be able to reference any properties on the agent and you'd assign them from anywhere that you could usually set state in your agent object, be that on creation or in agent functions that are invoked by mistreevous. I would say that setting state via a getter accessor might make the flow of processing a little complicated and hard to follow unless you are simply using it to initialise some state.

I think for now having dot or bracket accessors in our definitions may be a bit too much and would lead to complex definitions so I think I would stick with single property references for the sake of readability and to avoid having to dig through potentially complex types in our tree definitions.

I will get started with the variable references part of the work and will leave the subtree parameterisation bit for after as I need to have a think about the mdsl syntax for that.

Thanks again, please let me know if you have any more input 😄
